File: 0033-search-in-rotated-sorted-array/0033-search-in-rotated-sorted-array.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        start, end = 0, len(nums) - 1
        while start<=end:
            mid = (start + end) // 2
            if(nums[mid] == target):
                return mid
            if nums[start] <= nums[mid]:
                if target >=nums[start] and target < nums[mid]:
                    end = mid - 1
                else:
                    start = mid + 1
            elif target > nums[mid] and target <= nums[end]:
                start = mid + 1
            else:
                end = mid - 1
        return -1

File: 0033-search-in-rotated-sorted-array/0033-search-in-rotated-sorted-array.c
int search(int* nums, int numsSize, int target) {
    int start = 0;
    int end = numsSize - 1;
    
    while (start <= end) {
        int mid = start + (end - start) / 2;
        
        if (nums[mid] == target) {
            return mid;
        }
        
        if (nums[start] <= nums[mid]) {
            if (target >= nums[start] && target < nums[mid]) {
                end = mid - 1;
            } else {
                start = mid + 1;
            }
        } else {
            if (target > nums[mid] && target <= nums[end]) {
                start = mid + 1;
            } else {
                end = mid - 1;
            }
        }
    }
    
    return -1;
}